core/include/scipp/core/values_and_variances_strong_types.h

```diff
@@ -0,0 +1,138 @@
+  using OptionalContainer<T, Vector>::OptionalContainer;
+  template <class U> operator Values<U>() { return Values<U>(this->data); }
+};
+template <class T> Values(Vector<T> &&)->Values<T>;
+template <class T> Values(std::initializer_list<T>)->Values<T>;
+
+struct VariancesTag {};
+
+template <class T>
+struct Variances : VariancesTag, OptionalContainer<T, Vector> {
+  using OptionalContainer<T, Vector>::OptionalContainer;
+  template <class U> operator Variances<U>() {
+    return Variances<U>(this->data);
+  }
+};
+template <class T> Variances(Vector<T> &&)->Variances<T>;
+template <class T> Variances(std::initializer_list<T>)->Variances<T>;
+
+namespace detail {
+template <class T1, class T2>
+constexpr bool is_values_or_variances_v =
+    (std::is_base_of_v<ValuesTag, T1> && std::is_base_of_v<ValuesTag, T2>) ||
+    (std::is_base_of_v<VariancesTag, T1> &&
+     std::is_base_of_v<VariancesTag, T2>);
+
+template <class T1, class T2> struct is_same_or_values_or_variances {
+  static constexpr bool value =
+      std::is_same_v<T1, T2> || is_values_or_variances_v<T1, T2>;
+};
+template <class T1, class T2>
+bool constexpr is_same_or_values_or_variances_v =
+    is_same_or_values_or_variances<T1, T2>::value;
+
+template <class T, class... Args>
+constexpr bool is_type_in_pack_v = std::disjunction<
+    is_same_or_values_or_variances<T, std::decay_t<Args>>...>::value;
+
+template <class T, class... Args> struct Indexer {
+  template <std::size_t... IS>
+  static constexpr auto indexOfCorresponding_impl(std::index_sequence<IS...>) {
+    return ((is_same_or_values_or_variances_v<T, Args> * IS) + ...);
+  }
+
+  static constexpr auto indexOfCorresponding() {
+    return indexOfCorresponding_impl(
+        std::make_index_sequence<sizeof...(Args)>{});
+  }
+};
+
+template <class... Ts>
+// given types
+struct ConstructorArgumentsMatcher {
+  template <class... Args>
+  // needed types
+  constexpr static void checkArgTypesValid() {
+    static_assert((is_type_in_pack_v<Args, Ts...> + ...) == sizeof...(Ts));
+  }
+  template <class T, class... Args> static T construct(Ts &&... ts) {
+    auto tp = std::make_tuple(std::forward<Ts>(ts)...);
+    return T(std::forward<Args>(extractArgs<Args, Ts...>(tp))...);
+  }
+
+private:
+  template <class T, class... Args>
+  static decltype(auto) extractArgs(std::tuple<Args...> &tp) {
+    if constexpr (!is_type_in_pack_v<T, Ts...>)
+      return T{};
+    else {
+      constexpr auto index = Indexer<T, Args...>::indexOfCorresponding();
+      using Type = std::decay_t<decltype(std::get<index>(tp))>;
+      if constexpr (std::is_same_v<Type, T>)
+        return std::get<index>(tp);
+      else {
+        if constexpr (is_values_or_variances_v<T, Type>) {
+          using T1 = typename T::value_type;
+          using T2 = typename Type::value_type;
+          if constexpr (std::is_convertible_v<T1, T2>) {
+            return std::get<index>(tp);
```

Comment:
If you change this to something like:
```cpp
auto &&arg = std::get<index>(tp);
return Vector<T>(arg.begin(), arg.end());
```
... can you then remove `struct OptionalContainer` and just use `std::optional<Vector<T>>` in `Values` and `Variances`? Or at least, remove all the custom type cast operators and constructors.

Or even simpler, since the `element_array` stuff is merged, just use that? It does all the optional stuff, and maybe all we need is to add a couple of constructors?